### PR TITLE
Fix some warnings for GHC 8.0, 8.4 up, 9.4

### PR DESCRIPTION
**Pull request by andreasabel**
Fix some warnings for GHC 8.0, 8.4 up, 9.4, to be included in next release.

### CODE REVIEW

**Comment by andreasabel** [comment]
@andrewthad : PR is ready.

File: Data/Primitive/Array.hs [hunk 1]
```diff
@@ -762,7 +762,7 @@ instance Semigroup (Array a) where
 instance Monoid (Array a) where
   mempty = empty
 #if !(MIN_VERSION_base(4,11,0))
-  mappend = (<|>)
```
Comment (treeowl) [inline Data/Primitive/Array.hs:765]: Shouldn't the version guard be sufficient?

File: Data/Primitive/SmallArray.hs [hunk 2]
```diff
@@ -798,7 +795,7 @@ instance Sem.Semigroup (SmallArray a) where
 instance Monoid (SmallArray a) where
   mempty = empty
 #if !(MIN_VERSION_base(4,11,0))
-  mappend = (<|>)
+  mappend = (Sem.<>)
```
Comment (treeowl) [inline Data/Primitive/SmallArray.hs:798]: Again, why isn't the CPP enough?

File: Data/Primitive/Array.hs (hunk 1 above)
Comment (andreasabel) [reply Data/Primitive/Array.hs:765]: This change (to `(<>)` which is defined as `(<|>)`) is to silence the `noncanonical-mappend` warning under GHC 8.0 and 8.2 .

File: Data/Primitive/SmallArray.hs (hunk 2 above)
Comment (andreasabel) [reply Data/Primitive/SmallArray.hs:798]: See above.

**Comment by andrewthad** [comment]
I merged your other PR, but now there's a merge conflict in the cabal.project file. Can you resolve this, and then I can merge.

File: Data/Primitive/Array.hs (hunk 1 above)
Comment (andreasabel) [reply Data/Primitive/Array.hs:765]: Well, if you ask GHC 8.0.2, it is not happy about `mappend = (<|>)`.  
```shell-session
$ cabal build -w ghc-8.0.2
...
Data/Primitive/Array.hs:765:3: warning: [-Wnoncanonical-monoid-instances]
    Noncanonical ‘mappend’ definition detected
    in the instance declaration for ‘Monoid (Array a)’.
    Define as ‘mappend = (<>)’
```
If I lower the version in the CPP to `base-4.9.0` (which removes the `mappend` definition), GHC 8.0.2 isn't happy either:
```shell-session
$ cabal build -w ghc-8.0.2
...
Data/Primitive/Array.hs:762:10: warning: [-Wmissing-methods]
    • No explicit implementation for
        ‘mappend’
    • In the instance declaration for ‘Monoid (Array a)’
```
GHC 8.0.2 is like a defiant child, it wants its way, which is `mappend = (<>)`. Basta!

**Comment by andreasabel** [comment]
@andrewthad : I resolved the merge conflict.

File: Data/Primitive/Array.hs (hunk 1 above)
Comment (treeowl) [reply Data/Primitive/Array.hs:765]: Wow, that was lousy behavior indeed!

**Comment by andrewthad** [comment]
Merged. If you got anything else that you want to get in before I release 0.8.0.0, let me know. I'm trying to get that out to hackage soon.

**Comment by andreasabel** [comment]
Thanks!  
Atm, I cannot think of anything I'd like to change.

**Comment by andreasabel** [comment]
Actually, there are still 3 warning left for GHC 9.6:

Deprecation of `TypeInType`:
```
Data/Primitive/Types.hs:4:14: warning: [-Wdeprecated-flags]
    -XTypeInType is deprecated: use -XDataKinds and -XPolyKinds instead
  |
4 | {-# LANGUAGE TypeInType #-}
  |              ^^^^^^^^^^
```

2 unused imports due to new `Prelude`:
```
Data/Primitive/Types.hs:52:1: warning: [-Wunused-imports]
    The import of ‘Data.Ord’ is redundant
      except perhaps to import instances from ‘Data.Ord’
    To import instances alone, use: import Data.Ord()
   |
52 | import Data.Ord (Down(..))
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^

Data/Primitive/PrimArray.hs:116:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
    |
116 | import Control.Applicative
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

If you want to fix these...

Maybe the deprecation can be fixed without CPP.

The unused imports would need new CPP `#if`s, I suppose.

**Pull request merged**